### PR TITLE
KAFKA-8164: Improve test passing rate by rerunning flaky tests

### PR DESCRIPTION
**Pull request by viktorsomogyi**
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by viktorsomogyi** [comment]
@ijuma I've created the new PR about flaky tests. Currently it lets 3 retries for a test and fails the build after 50 test failure. Both numbers are a bit arbitrary (especially the 50 :) ), but if you know a good constant for them, I'd be happy to change them.

Furthermore I ran into a few cases where the `WhateverClass.classMethod` failed which usually happens if there were unexpected running threads after the test. These obviously can't be retried and the plugin will fail as well (and subsequently fails the build too):

```
* What went wrong:
Execution failed for task ':core:test'.
> org.gradle.test-retry was unable to retry the following test methods, which is unexpected. Please file a bug report at https://github.com/gradle/test-retry-gradle-plugin/issues
     kafka.api.MetricsTest#classMethod
     kafka.api.PlaintextAdminIntegrationTest#classMethod
     kafka.api.SslAdminIntegrationTest#classMethod
     kafka.api.ConsumerTopicCreationTest#classMethod
     kafka.api.SaslSslAdminIntegrationTest#classMethod
```
Not sure if this is a big problem but I wanted to let you know. The old custom flaky retry didn't handle these either.

File: build.gradle [hunk 1]
```diff
@@ -289,6 +291,11 @@ subprojects {
       exceptionFormat = testExceptionFormat
     }
     logTestStdout.rehydrate(delegate, owner, this)()
+
+    retry {
+      maxRetries = 3
+      maxFailures = 50
```
Comment (ijuma) [inline build.gradle]: I would go with something conservative: 1 retry and max 5 failures.We want to fix flaky tests, but if something just fails once, it's not worth failing the build given Jenkins flakiness.

Comment (ijuma) [reply build.gradle]: We can adjust after we see how it works in practice, but this would be a good starting point IMO.

Comment (viktorsomogyi) [reply build.gradle]: Ok. I figured we can also make these configurable. Updated the PR.

File: build.gradle [hunk 2]
```diff
@@ -97,6 +98,9 @@ ext {
 
   userMaxForks = project.hasProperty('maxParallelForks') ? maxParallelForks.toInteger() : null
 
+  userFlakyRetries = project.hasProperty('flakyRetries') ? flakyRetries.toInteger() : 1
+  userMaxTestFailures = project.hasProperty('maxTestFailures') ? maxTestFailures.toInteger() : 5
```
Comment (ijuma) [inline build.gradle]: How about maxTestRetries and maxTestRetryFailures? Also, we need to update the Readme to mention these.

Comment (viktorsomogyi) [reply build.gradle]: Ok

**Comment by viktorsomogyi** [comment]
Updated the common configs section

**Comment by ijuma** [comment]
ok to test

File: gradle/dependencies.gradle [hunk 3]
```diff
@@ -112,6 +112,7 @@ versions += [
   spotbugs: "3.1.12",
   spotbugsPlugin: "3.0.0",
   spotlessPlugin: "3.27.1",
+  testRetryPlugin: "1.0.2",
```
Comment (chia7712) [inline gradle/dependencies.gradle]: How about using latest version (1.1.0)?

Comment (ijuma) [reply gradle/dependencies.gradle]: I also noticed it was released shortly after this PR was submitted. Updated.

**Comment by ijuma** [comment]
ok to test

**Review by ijuma — APPROVED** [review]
LGTM

**Comment by ijuma** [comment]
@viktorsomogyi I made some minor changes, please check that you're happy with them.

**Comment by ijuma** [comment]
ok to test

**Comment by ijuma** [comment]
retest this please

**Comment by ijuma** [comment]
Thinking about this some more and after some discussion, I think we should probably default to 0 retries and tweak the PR builder to pass 1 and 5. Will update the PR if @viktorsomogyi doesn't beat me to it.

**Comment by viktorsomogyi** [comment]
@ijuma I was behind you 10 minutes or so, waited for the local build to run and I saw it only after that you already pushed. :)
I'm happy with the changes.

**Comment by ijuma** [comment]
retest this please

**Comment by ijuma** [comment]
Updated Jenkins PR jobs to be:

> ./jenkins.sh -PmaxParallelForks=2 -PscalaVersion=2.12 -PmaxTestRetries=1 -PmaxTestRetryFailures=5

Branch builds have not been changed.

**Comment by ijuma** [comment]
One flaky failure with retries (2.12 job):

> kafka.admin.ResetConsumerGroupOffsetTest.testResetOffsetsExportImportPlan

Three flaky failures without retries (2.13 job):

> kafka.admin.DescribeConsumerGroupTest.testDescribeGroupWithShortInitializationTimeout
> org.apache.kafka.streams.kstream.internals.KStreamImplTest.shouldSupportTriggerMaterializedWithKTableFromKStream
> org.apache.kafka.streams.processor.internals.StoreChangelogReaderTest.shouldRequestCommittedOffsetsAndHandleTimeoutException[1]

I'll go ahead and merge this so that we can evaluate how well it works in practice.

**Comment by ijuma** [comment]
Thanks for the contribution!

**Comment by viktorsomogyi** [comment]
@ijuma thank you for reviewing this change!

File: README.md [hunk 4]
```diff
@@ -48,6 +48,13 @@ Change the log4j setting in either `clients/src/test/resources/log4j.properties`
 
     ./gradlew clients:test --tests RequestResponseTest
 
+### Specifying test retries ###
+By default, each failed test is retried once up to a maximum of five retries per test run. Tests are retried at the end of the test task. Adjust these parameters in the following way:
```
Comment (akatona84) [inline README.md:52]: Maybe we should update this line that by default it is not retrying.

**Pull request merged**